Grant/contracts/BuilderGrant.sol

```diff
@@ -0,0 +1,350 @@
+
+    function currentSeason() external view returns (uint256) {
+        return seasons.length; // The last claimable season = length - 1
+    }
+
+    function deposit(uint256 huntAmount) external {
+        address msgSender = _msgSender();
+        if (!HUNT.transferFrom(msgSender, address(this), huntAmount)) revert TokenTransferFailed();
+
+        emit Deposit(msgSender, huntAmount);
+    }
+
+    function emergencyWithdraw() external onlyOwner {
+        address msgSender = _msgSender();
+        uint256 balance = HUNT.balanceOf(address(this));
+        if (balance == 0) revert NothingToWithdraw();
+        if (!HUNT.transfer(msgSender, balance)) revert TokenTransferFailed();
+
+        emit EmergencyWithdraw(msgSender, balance);
+    }
+
+    function setSeasonData(
+        uint256 seasonId,
+        uint16[3] calldata grantsAmount,
+        uint48[] calldata fids,
+        address[] calldata wallets
+    ) external onlyOwner {
+        // Validate params
+        if (fids.length == 0 || fids.length != wallets.length) revert InvalidRankersParams();
+
+        // Check if we have enough HUNT to mint all Mini Building NFTs
+        if (
+            HUNT.balanceOf(address(this)) <
+            HUNT_PER_MINI_BUILDING * (grantsAmount[0] + grantsAmount[1] + grantsAmount[2])
+        ) revert NotEnoughGrantBalance();
+
+        // Check if there are enough ranker data is provided if the 1st ranker donates 100%
+        // e.g. If the top grant is 10,000 HUNT, include up to 103 rankers to allow donations to ranks 4-103
+        if (grantsAmount[0] > (fids.length - 3)) revert InvalidGrantAmount();
+
+        // Check if the grants amount is even because they have 50% donation option
+        if (grantsAmount[0] % 2 != 0 || grantsAmount[1] % 2 != 0 || grantsAmount[2] % 2 != 0)
+            revert InvalidGrantAmount();
+
+        Season storage season;
+        // Overwriting the last season
+        if (seasons.length > 0 && seasonId == seasons.length - 1) {
+            season = seasons[seasonId];
+            // can't overwrite the data of the season if anyone has claimed
+            if (season.totalClaimed > 0) revert SeasonDataIsNotUpdateable();
+            // Set new season
+        } else if (seasonId == seasons.length) {
+            seasons.push();
+            season = seasons[seasonId];
+        } else {
+            revert InvalidSeasonId();
+        }
+
+        // Set grant data
+        season.grants[0].amount = grantsAmount[0];
+        season.grants[1].amount = grantsAmount[1];
+        season.grants[2].amount = grantsAmount[2];
+        season.claimStartedAt = uint40(block.timestamp);
+
+        // Reset ranker data if exists, and overwrites the old data
+        if (season.rankers.length > 0) {
+            delete season.rankers;
+        }
+        unchecked {
+            for (uint256 i = 0; i < fids.length; ++i) {
+                season.rankers.push(Ranker(fids[i], wallets[i], 0, [false, false, false]));
+            }
```

Comment:
GAS: using a memory array of defaults reduces gas from avg `651,285` to `650,304`

Not particularly significant but it was worth a try on this expensive function.

```solidity
    bool[3] memory defaults = [false, false, false];
    for (uint256 i = 0; i < fids.length; ++i) {
        season.rankers.push(Ranker(fids[i], wallets[i], 0, defaults));
    }
```solidity

Reply:
Great! I love these types of little gas-saving tricks 👍